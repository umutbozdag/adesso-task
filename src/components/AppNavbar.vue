<template>
  <nav class="navbar bg-black border-gray-200 mb-12">
    <div class="flex flex-wrap items-center justify-between mx-auto py-4">
      <a class="flex items-center">
        <span class="self-center text-2xl font-semibold text-white">App</span>
      </a>
      <ul
        class="font-medium flex rounded-lg bg-black flex-row md:space-x-8 rtl:space-x-reverse mt-0 border-0 border-gray-700"
      >
        <template v-for="route in routes" :key="route.name">
          <li class="navbar__item" v-if="route.showInNavbar">
            <router-link
              :to="{ name: route.name }"
              class="navbar__link rounded md:bg-transparent px-2 py-1 text-white"
              >{{ route.label }}</router-link
            >
          </li>
        </template>
      </ul>
    </div>
  </nav>
</template>

<script setup lang="ts">
import routes from '@/config/routes';
</script>

<style lang="scss" scoped>
.navbar__link {
  &:hover {
    color: $color-beige;
    transition: 0.2s ease;
  }
}

.router-link-exact-active {
  background-color: $active-route-bg;
  color: $color-beige;
}
</style>
