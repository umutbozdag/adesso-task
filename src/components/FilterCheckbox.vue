<template>
  <div class="flex items-center mr-1">
    <input
      type="checkbox"
      :checked="modelValue"
      :value="modelValue"
      @change="toggleCheckbox"
      class="w-4 h-4 rounded focus:ring-blue-600 ring-offset-gray-800 focus:ring-2 bg-gray-700 border-gray-600"
    />
    <label v-if="label" class="text-sm font-medium text-gray-300 mr-2">{{ label }}</label>
  </div>
</template>

<script setup lang="ts">
interface IFilterCheckboxProps {
  modelValue: boolean;
  label?: string;
}

withDefaults(defineProps<IFilterCheckboxProps>(), {
  modelValue: false,
  label: ''
});

const emit = defineEmits<{
  (e: 'update:modelValue', isChecked: boolean): void;
}>();

const toggleCheckbox = (e: Event) => {
  emit('update:modelValue', (e.target as HTMLInputElement).checked);
};
</script>
