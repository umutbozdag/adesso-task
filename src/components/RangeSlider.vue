<template>
  <div class="flex items-center">
    <slot></slot>
    <label class="text-sm font-medium text-white mr-2">
      <slot name="slider-label"></slot>
    </label>
    <input
      type="range"
      :value="modelValue"
      :min="min"
      :max="max"
      :disabled="disabled"
      @change="handleChange"
      class="w-full h-2 rounded-lg appearance-none cursor-pointer bg-gray-700"
    />
  </div>
  <span v-if="modelValue !== '0'" class="text-white ml-2 mr-8"
    >{{ MIN_DEFAULT }} - {{ modelValue }}</span
  >
</template>

<script setup lang="ts">
interface IRangeSliderProps {
  min?: number;
  max?: number;
  modelValue: string;
  disabled?: boolean;
}
const emit = defineEmits<{
  (e: 'update:modelValue', id: string): void;
}>();

const MIN_DEFAULT = 0;

withDefaults(defineProps<IRangeSliderProps>(), {
  min: 0,
  max: 200,
  modelValue: '0',
  disabled: false
});

const handleChange = (e: Event) => {
  emit('update:modelValue', (e.target as HTMLInputElement).value);
};
</script>
