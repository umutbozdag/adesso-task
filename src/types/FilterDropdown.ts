export interface FilterDropdownOption {
  label: string;
  name: string;
}

export enum FilterDropdownEmitEnum {
  OPTION_SELECTED = 'option-selected'
}
