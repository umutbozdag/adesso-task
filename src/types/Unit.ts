export interface IUnits {
  units: IUnit[];
}

export interface IUnit {
  id: number;
  name: string;
  description: string;
  expansion: UnitExpansionType;
  age: UnitAgeType;
  cost: IUnitCost | null;
  build_time?: number;
  reload_time?: number;
  attack_delay?: number;
  movement_rate?: number;
  line_of_sight: number;
  hit_points: number;
  range?: UnitRangeType;
  attack?: number;
  armor: string;
  accuracy?: string;
  attack_bonus?: string[];
  search_radius?: number;
  blast_radius?: number;
  armor_bonus?: string[];
}

export enum UnitAgeEnum {
  All = 'All',
  Feudal = 'Feudal',
  Castle = 'Castle',
  Imperial = 'Imperial',
  Dark = 'Dark'
}

export type UnitAgeType = Exclude<UnitAgeEnum, UnitAgeEnum.All>;

export interface IUnitCost {
  Wood?: number;
  Gold?: number;
  Food?: number;
}

export type UnitExpansionType = 'Age of Kings' | 'The Conquerors';

export type UnitRangeType = number | string;
