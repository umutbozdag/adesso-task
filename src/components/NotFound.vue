<template>
  <div class="flex items-center justify-center h-64">
    <div class="text-center">
      <h1 class="text-6xl font-bold text-gray-800">404</h1>
      <p class="text-2xl font-semibold text-gray-600 mb-4">Not Found</p>
      <router-link :to="{ name: RoutesNameEnum.Home }" class="text-blue-500 hover:underline"
        >Go back to Home</router-link
      >
    </div>
  </div>
</template>

<script setup lang="ts">
import { RoutesNameEnum } from '@/types/Route';
</script>
