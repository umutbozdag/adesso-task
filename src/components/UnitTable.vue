<template>
  <div class="unit-table relative py-8">
    <div class="unit-table__wrapper overflow-y-auto">
      <table class="text-sm text-left text-gray-500 w-full table-auto">
        <thead class="text-xs uppercase bg-gray-900 text-gray-200 sticky top-0">
          <tr>
            <unit-table-header-item
              v-for="tableHeader in tableHeaders"
              :key="tableHeader.id"
              :table-header="tableHeader"
            />
          </tr>
        </thead>
        <tbody class="unit-table__rows h-96">
          <unit-table-row-item
            v-for="(unit, idx) in units"
            :key="unit.id"
            :row-index="idx"
            :unit-detail="unit"
            @row-click="handleRowClick"
          />
        </tbody>
      </table>
    </div>
  </div>
</template>

<script setup lang="ts">
import { IUnitDTO } from '@/types/dto/UnitDTO';
import UnitTableRowItem from './UnitTableRowItem.vue';
import UnitTableHeaderItem from './UnitTableHeaderItem.vue';
import { useRouter } from 'vue-router';
import { RoutesNameEnum } from '@/types/Route';
import { IUnitTableHeaderItem } from '@/types/UnitTable';

interface IUnitTableProps {
  units: IUnitDTO[];
  tableHeaders: IUnitTableHeaderItem[];
}

defineProps<IUnitTableProps>();

const router = useRouter();

const handleRowClick = (item: IUnitDTO) => {
  router.push({
    name: RoutesNameEnum.UnitDetails,
    params: {
      id: item.id
    }
  });
};
</script>

<style lang="scss" scoped>
.unit-table {
  &__wrapper {
    max-height: $table-max-height;
  }
}
</style>
