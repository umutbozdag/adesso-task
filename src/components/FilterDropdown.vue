<template>
  <div class="z-10 inline-block" v-on-click-outside="closeDropdown">
    <button
      @click="isOpen = !isOpen"
      class="justify-center w-full px-4 py-2 text-sm font-medium text-white bg-gray-900 border border-gray-600 rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-offset-gray-100 focus:ring-blue-500"
    >
      {{ selectedOptionLabel }}: {{ selectedOption }}
    </button>
    <div
      v-if="isOpen"
      class="origin-top-right absolute mt-2 w-48 rounded-md shadow-lg z-10 bg-gray-900 ring-1 ring-black ring-opacity-5"
    >
      <div class="py-2 p-2" role="menu">
        <a
          v-for="option in options"
          :key="option.name"
          class="flex rounded-md px-4 py-2 text-sm text-white hover:bg-gray-800 cursor-pointer"
          role="menuitem"
          @click="handleOptionClick(option)"
        >
          {{ option.label }}
        </a>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { UnitAgeEnum } from '@/types/Unit';
import { FilterDropdownOption, FilterDropdownEmitEnum } from '@/types/FilterDropdown';
import { ref } from 'vue';
import { vOnClickOutside } from '@vueuse/components';

interface IFilterDropdownProps {
  options: FilterDropdownOption[];
  selectedOption: UnitAgeEnum;
  selectedOptionLabel: string;
}

defineProps<IFilterDropdownProps>();

const emit = defineEmits<{
  (event: 'option-selected', option: FilterDropdownOption): void;
}>();

const isOpen = ref(false);

const closeDropdown = () => {
  isOpen.value = false;
};

const handleOptionClick = (option: FilterDropdownOption) => {
  emit(FilterDropdownEmitEnum.OPTION_SELECTED, option);
  closeDropdown();
};
</script>
@/types/FilterDropdown
