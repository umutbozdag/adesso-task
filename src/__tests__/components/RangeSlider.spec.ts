import FilterDropdown from '@/components/FilterDropdown.vue';
import RangeSlider from '@/components/RangeSlider.vue';
import { UnitAgeEnum } from '@/types/Unit';
import { VueWrapper, mount } from '@vue/test-utils';
import { OnClickOutside } from '@vueuse/components';
import { expect, it, describe, beforeAll } from 'vitest';
import { ComponentPublicInstance } from 'vue';

describe('RangeSlider', () => {
  let wrapper: VueWrapper<ComponentPublicInstance & { [key: string]: any }>;

  beforeAll(() => {
    wrapper = mount(RangeSlider, {
      props: {
        modelValue: '0'
      }
    });
  });

  it('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('handleChange emits event', async () => {
    const input = document.createElement('input');
    input.value = 'input';
    const event = new Event('input');
    input.dispatchEvent(event);
    wrapper.vm.handleChange(event);
    const emittedEvents = wrapper.emitted();
    expect(emittedEvents['update:modelValue']).toHaveLength(1);
    expect(emittedEvents['update:modelValue'][0]).toEqual(['input']);
  });
});
