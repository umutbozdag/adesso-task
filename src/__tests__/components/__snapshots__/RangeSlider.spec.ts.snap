// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RangeSlider > renders correctly 1`] = `
"<div class=\\"flex items-center\\"><label class=\\"text-sm font-medium text-white mr-2\\"></label><input type=\\"range\\" min=\\"0\\" max=\\"200\\" class=\\"w-full h-2 rounded-lg appearance-none cursor-pointer bg-gray-700\\"></div>
<!--v-if-->"
`;
